src/lib/Microsoft.Health.Events/EventConsumers/Service/Infrastructure/EventQueueWindow.cs

```diff
@@ -0,0 +1,34 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+
+namespace Microsoft.Health.Events.EventConsumers.Service.Infrastructure
+{
+    public class EventQueueWindow
+    {
+        private DateTime _windowEnd = DateTime.MinValue;
+        private TimeSpan _flushTimespan;
+
+        public EventQueueWindow(DateTime initDateTime, TimeSpan flushTimespan)
+        {
+            _windowEnd = initDateTime.Add(flushTimespan);
+            _flushTimespan = flushTimespan;
+        }
+
+        public void IncrementWindow(DateTime currentEnqueudTime)
+        {
+            while (currentEnqueudTime >= _windowEnd)
+            {
+                _windowEnd = _windowEnd.Add(_flushTimespan);
```

Comment:
You should be able to do this with math.  The difference between the times will result in a timespan.  You should be able to determine how many times bigger the time is from the window end.  You can then scale the window end by X amount of the timespan.